source/extensions/common/wasm/context.cc

```diff
@@ -1640,10 +1629,8 @@ WasmResult Context::sendLocalResponse(uint32_t response_code, std::string_view b
       }
       decoder_callbacks_->sendLocalReply(static_cast<Envoy::Http::Code>(response_code), body_text,
```

Comment:
@mpwarres The spec is a bit unclear about the semantics of send_local_reply. It looks like it is delayed in Envoy implementation, e.g. nothing happens until after the callback completes, and then response callbacks are invoked later. Is this consistent with your thinking?

Reply:
FWIW, I think your understanding is correct, it's delayed in Envoy (actually in two places for some code paths, here in the context.cc and in prepareLocalReplyViaFilterChain). It was the behavior ever since proxy-wasm was upstreamed (see https://github.com/envoyproxy/envoy/commit/319a9a647f19036d831a75f68350c90e98bfc365).

Calling send_local_reply synchronously may cause re-entering proxy-wasm plugin, which does not seem to be a safe thing to do generally (even if some plugins may be fine with that), though I cannot come up with any specific example where it would cause an issue from the top of my head.

Reply:
I'll follow up with a spec update because it is ambiguous at it is specified.